tc/tc_util.c

```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
 };
 
 
+int get_percent_rate(unsigned int *rate, const char *str, char *dev)
+{
+	long value;
+	char percent[strlen(dev) - 1], r_str[20];
+	const char *pos;
+	int ret;
+	double d, r;
+
+	if (read_prop(dev, "speed", &value))
+		return -1;
+
+	pos = str + strlen(str) -1;
+	if (*pos != '%') {
+		fprintf(stderr, "Specified rate value is malformed \n");
+		return -1;
+	}
+
+	strncpy(percent, str, strlen(str) -1);
+	ret = sscanf(percent, "%lf", &d);
```

Comment:
Actually, instead of trying to check whether `str` ends with a `%` and copying everything except the `%` into the variable `percent`, why not just make this statement:

    ret = sscanf(percent, "%lf%%", &d);

Then `sscanf()` will be expecting a floating point number followed by a percentage character.

Reply:
Then we can get rid of the string `percent`, pointer `pos`, and the `if` statement  above.

Reply:
Done. Works good.